### PR TITLE
feature/39_COMMENT_FUNCTIONALITY

### PR DESCRIPTION
**Pull request by Symphonie234**
### Changes Made

- Added a functionality to add comments
- Added a functionality to like comments
- Added a new **Controller** file named **`CommentController`**
- Added **`Comment`** and **`CommentLike`** Models
- Added new routes in the **`web.php`** route file
- Added **migration** files for comments and likes on it

### CODE REVIEW

File: app/Http/Controllers/CommentController.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * Store a new comment.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\RedirectResponse
+     */
+    public function store(Request $request)
+    {
+        // Validate the request data
+        $validatedData = $request->validate([
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php]: make validation file

File: app/Http/Controllers/CommentController.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            'content' => 'required|string',
+            // Add any other validation rules as needed
+        ]);
+
+        // Create a new comment in the database
+        $comment = new Comment([
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php]: Please use register trait

File: app/Http/Controllers/CommentController.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            'user_id' => auth()->user()->id, // Associate the comment with the authenticated user
+        ]);
+
+        $comment->save();
+
+        // Broadcast the new comment
+        broadcast(new CommentPosted($comment))->toOthers();
+
+        // Redirect back to the specific listing page
+        return redirect()->route('listings.show', ['listing' => $request->input('listing_id')])
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php]: use message lang

File: app/Http/Controllers/CommentController.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * Retrieve all comments for a listing.
+     *
+     * @param  int  $listingId
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function index($listingId)
+    {
+        // Replace 'Comment' with your actual Comment model
+        $comments = Comment::where('listing_id', $listingId)->latest()->get();
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php:74]: don't have return?

File: app/Http/Controllers/CommentController.php [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * Like a comment and reload the page.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  int  $commentId
+     * @return \Illuminate\Http\RedirectResponse
+     */
+    public function like(Request $request, $commentId)
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php]: $request parameter looks like not use

File: app/Http/Controllers/CommentController.php [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            // User has already liked the comment, remove the like
+            $commentLike = CommentLike::where('comment_id', $comment->id)
+                ->where('user_id', $userId)
+                ->first();
+
+            if ($commentLike) {
+                $commentLike->delete();
+            }
+
+            return redirect()->back()->with('success', 'Comment unliked successfully');
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php]: use message lang

File: app/Http/Controllers/CommentController.php [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            $commentLike = CommentLike::where('comment_id', $comment->id)
+                ->where('user_id', $userId)
+                ->first();
+
+            if ($commentLike) {
+                $commentLike->delete();
+            }
+        }
+
+        // Redirect back to the page (the comment's parent listing) after unliking
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php]: use message lang

File: app/Http/Controllers/ListingController.php [hunk 8]
```diff
@@ -199,4 +203,32 @@ public function unlike(Listing $listing)
...
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function storeComment(Request $request)
+    {
+        $user = auth()->user();
+
+        // Validate the request data for the comment
+        $validatedData = $request->validate([
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: make validation file

File: app/Http/Controllers/ListingController.php [hunk 9]
```diff
@@ -199,4 +203,32 @@ public function unlike(Listing $listing)
...
+            'listing_id' => 'required|exists:listings,id',
+            'content' => 'required|string|max:255',
+        ]);
+
+        // Create a new comment
+        $comment = new Comment([
```
Comment (ChronoDevs) [inline app/Http/Controllers/ListingController.php]: use register trait

**Comment by ChronoDevs** [comment]
Please fix the conflict. Thank you

**Comment by Symphonie234** [comment]
### Changes that have been addressed in the PR No. 1:

- Created two new **Requests** files named **`CommentStoreRequest.php`** and **`ListingStoreCommentRequest.php`**.
- Have the **`store`** method in the **CommentController** and **`storeComment`** method in the **`ListingController`** use the **`Register.php`** **Trait**.
- Created new messages in the **`lang/messages.php`** to be used in the **`CommentController`** methods.
- Fixed the conflict in the **`User`** **Model**.

**Comment by Symphonie234** [comment]
### New Change/s:

- Added a feature to show the number of comments the listing has in the **`listings/show.blade.php`**.

File: app/Events/CommentPosted.php [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+use Illuminate\Foundation\Events\Dispatchable;
+use Illuminate\Queue\SerializesModels;
+
+class CommentPosted implements ShouldBroadcast
+{
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    public $comment;
+
+    public function __construct($comment)
```
Comment (ChronoDevs) [inline app/Events/CommentPosted.php:25]: Please add comment method above every function. Thank you

File: app/Http/Controllers/CommentController.php [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+use Illuminate\Support\Facades\Auth;
+use App\Events\CommentPosted;
+use App\Models\Comment;
+use App\Models\CommentLike;
+use App\Models\Listing;
+use App\Http\Requests\CommentStoreRequest;
+
+class CommentController extends Controller
+{
+    public function __construct()
```
Comment (ChronoDevs) [inline app/Http/Controllers/CommentController.php:20]: Please add comment method above every function. Thank you

**Comment by Symphonie234** [comment]
### Changes that have been addressed in the PR No. 2:

- Added comments above the functions of the **Event** file **`CommentPosted.php`**.
- Added comments above the functions of the **Controller** file **`CommentController.php`**.

**Pull request merged**